src/zope/tales/tales.py

```diff
@@ -744,12 +744,7 @@ def setGlobal(self, name, value):
             vars[name] = value
 
     def getValue(self, name, default=None):
-        value = default
-        for vars in self._vars_stack:
-            value = vars.get(name, default)
-            if value is not default:
-                break
-        return value
+        return self.vars.get(name, default)
```

Comment:
The interface this object implements is defined in `zope.tal`. That interface doesn't say very much about `getValue`, but my instinct, based on the existence of `setGlobal` and `setLocal` is that `getValue` should examine the entire stack of scopes, not just the current scope. It seems to make sense to do what this was doing, just iterate from current to oldest instead of the reverse.

I think code in `zope.tal` supports this. For example, [here](https://github.com/zopefoundation/zope.tal/blob/f3a8f2007361800effb4b730cad4327b78ffe43e/src/zope/tal/talinterpreter.py#L895) it appears to be wanting any current value before shadowing it in the current scope:
```python
        # We want 'macroname' name to be always available as a variable
        outer = self.engine.getValue('macroname')
        self.engine.setLocal('macroname', macroName.rsplit('/', 1)[-1])
```

I think the tests down below also basically support this interpretation, as they access shadowed variables and get values from outer scopes.

But they do that directly using `.vars`, not `getValue`. They only reason they work is because `beginScope` starts with a copy of the current `.vars`. Since `setGlobal` modifies every scope on the stack, I think this works out OK. Presumably this (directly using `.vars`, and it being a copy) was done for variable lookup speed, although it makes nested scopes more expensive.

If that makes sense, a comment explaining why there's no need to iterate would be appreciated and save having to do this analysis in the future.

Aside
====
I thought for a minute that the bug here in `getValue` may be hidden by another bug in `zope.tal`, the handling of compiled expressions that access `setGlobal` [appears to be questionable](https://github.com/zopefoundation/zope.tal/blob/f3a8f2007361800effb4b730cad4327b78ffe43e/src/zope/tal/talinterpreter.py#L582-L593):
```python
    def do_setLocal(self, notused):
        pass
    def do_setLocal_tal(self, stuff):
        (name, expr) = stuff
        self.engine.setLocal(name, self.engine.evaluateValue(expr))
    bytecode_handlers["setLocal"] = do_setLocal

    def do_setGlobal_tal(self, stuff):
        (name, expr) = stuff
        self.engine.setGlobal(name, self.engine.evaluateValue(expr))
    bytecode_handlers["setGlobal"] = do_setLocal
```

but it turns out a different set of bytecode handlers [are used for tal expressions](https://github.com/zopefoundation/zope.tal/blob/f3a8f2007361800effb4b730cad4327b78ffe43e/src/zope/tal/talinterpreter.py#L1007-L1008):
```python
    bytecode_handlers_tal["setLocal"] = do_setLocal_tal
    bytecode_handlers_tal["setGlobal"] = do_setGlobal_tal
```